cpp/src/distance/neighbors/ivfpq_build.cu

```diff
@@ -19,43 +19,36 @@
-             uint32_t dim,                                                                   \
-             raft::neighbors::ivf_pq::index<IdxT>* idx)                                      \
-  {                                                                                          \
-    *idx = raft::neighbors::ivf_pq::build<T, IdxT>(handle, params, dataset, n_rows, dim);    \
-  }                                                                                          \
-  void extend(raft::device_resources const& handle,                                          \
-              raft::neighbors::ivf_pq::index<IdxT>* idx,                                     \
-              const T* new_vectors,                                                          \
-              const IdxT* new_indices,                                                       \
-              IdxT n_rows)                                                                   \
-  {                                                                                          \
-    raft::neighbors::ivf_pq::extend<T, IdxT>(handle, idx, new_vectors, new_indices, n_rows); \
+#define RAFT_INST_BUILD_EXTEND(T, IdxT)                                                            \
+  auto build(raft::device_resources const& handle,                                                 \
+             const raft::neighbors::ivf_pq::index_params& params,                                  \
+             const raft::device_matrix_view<const T, IdxT, row_major>& dataset)                    \
+    ->raft::neighbors::ivf_pq::index<IdxT>                                                         \
+  {                                                                                                \
+    return raft::neighbors::ivf_pq::build<T, IdxT>(handle, params, dataset);                       \
+  }                                                                                                \
+  auto extend(raft::device_resources const& handle,                                                \
+              const raft::neighbors::ivf_pq::index<IdxT>& orig_index,                              \
+              const raft::device_matrix_view<const T, IdxT, row_major>& new_vectors,               \
+              const raft::device_matrix_view<const IdxT, IdxT, row_major>& new_indices)            \
+    ->raft::neighbors::ivf_pq::index<IdxT>                                                         \
+  {                                                                                                \
+    return raft::neighbors::ivf_pq::extend<T, IdxT>(handle, orig_index, new_vectors, new_indices); \
+  }                                                                                                \
+                                                                                                   \
+  void build(raft::device_resources const& handle,                                                 \
```

Comment:
Yes! This is perfect (no pointer-based APIs in runtime, though they should still be removed from the header).